automata/fa/dfa.py

```diff
@@ -211,10 +223,75 @@ def __iter__(self) -> Iterator[str]:
             yield from self.words_of_length(i)
             i += 1
 
+    # Supports partial
     def __len__(self) -> int:
         """Returns the cardinality of the language represented by the DFA."""
         return self.cardinality()
 
+    # Supports partial (test)
+    def as_partial(self) -> DFA:
+        """
+        Turns a DFA (complete or not) into a partial DFA.
+        Removes dead states and trap states (except the initial state)
+            and all edges leading to them.
+        """
+        if self.allow_partial:
+            return self.copy()
+
+        graph = self._get_digraph()
```

Comment:
Now that I'm thinking of it, the digraph associated with a DFA should probably be made into a cached property.

Reply:
It is funny because I was going to propose quite the opposite. Getting rid of any cached property

Reply:
We don't need to discuss it here, but my problem is specifically with the word cache. Caching all words of length k is a (silent) memory bomb. In my opinion, this should be implemented as an iterator and if the users want fast access to this set, they should compute and store it themselves.

Caching digraph and word lengths seem to offer an acceptable memory overhead (I would still vote for its removal otherwise we need to worry about keeping it consistent)

Reply:
I do share @EduardoGoulart1's concern about bloating memory usage, but am not sure how to balance it if the alternative comes with a considerable performance impact to many methods. Maybe we make this a global configuration option (i.e. in [`automata.base.config`](https://github.com/caleb531/automata/blob/main/automata/base/config.py)) and leave it up to the user?

I'm a bit undecided on the right path forward, so @eliotwrobson, please let me know your thoughts, and perhaps we can all come to an agreement on the right approach.

Reply:
I think either a global or local config is the way to go. Kinda like the immutability changes, for most situations, the current memory usage isn't a huge deal, and it makes common operations run quickly (I think that that the accepted words are actually computed using dynamic programming, so memory usage at some level is unavoidable). But I agree that it would be nice to have the option to turn these off if people are running into performance issues. Setting this via global or local config should be fairly straightforward.

I don't think this quite applies yet, but if people are running into more and more performance issues, we could consider introducing Cython as a backend for this project. That will _substantially_ improve things on both of these fronts. However, this would be a significant effort, so maybe something to consider in v9 if this is still a concern.

Reply:
Okay, great! Since we have a couple different scenarios:

1. Cache words
2. Cache diggraph / word lengths
3. Disabled cache

I wonder if we should have some config property like `cache_mode` or `word_cache_mode` (@eliotwrobson, maybe help me with a good name here 😅), with an enum as the set of possible values.

As for the Cython thing, I have zero experience writing C extensions for Python. It does sound like a very interesting idea though, provided we have someone who knows what they're doing (not sure how familiar you yourself are).

Reply:
> Okay, great! Since we have a couple different scenarios:
> 
> 1. Cache words
> 2. Cache diggraph / word lengths
> 3. Disabled cache

@caleb531 maybe call this a `cache_level` enum, where higher levels equate to more caching (so cache everything > just cache the digraph > cache nothing). I apologize somewhat for derailing discussion here a bit, it's probably best to handle this after everything else for v8 has already been merged (since the code changes are going to be short and this is a performance improvement and not a new feature).

> As for the Cython thing, I have zero experience writing C extensions for Python. It does sound like a very interesting idea though, provided we have someone who knows what they're doing (not sure how familiar you yourself are).

I'm not really familiar with this either (although I actually think it's fairly straightforward, the code I've seen just looks like Python with some extra information). I think I mentioned this somewhere else, but most of the current implementations have really hit the limit of performance in terms of what can be done in pure Python. Everything here is efficient enough for my purposes, so I think this can wait if/when someone comes along needing to really push the efficiency and wants to take this on.

@EduardoGoulart1 speaking of which, if you find yourself needing minimization for DFAs over large alphabets, there actually is a specialized algorithm that performs better than the one implemented here. So even for this use case, we don't have to resort to introducing new technologies quite yet 😅 .

Reply:
Not sure about how the Cython extensions could look like, but I would prefer to avoid it as much as possible... 
@eliotwrobson which algorithm are you talking about? I only know [this one](https://www.sciencedirect.com/science/article/abs/pii/S0020019011003267)

Reply:
Let's bring this discussion somewhere else, I think that for now we can just go along with this PR (I'll try to wrap it up the next days)